New JMS ActiveMQ example for our examples/tutorials

sanjaypujare opened the pull request:
@amberarrow  @ashwinchandrap  pls review and commit.

Commit:
New JMS ActiveMQ example for our examples/tutorials
remove duplicate jdbcIngest line

amberarrow commented on tutorials/jmsActiveMQ/pom.xml:
```diff
@@ -0,0 +1,291 @@
+
+  <!-- change these to the appropriate values -->
+  <name>My Apex Application</name>
+  <description>My Apex Application Description</description>
```
Suggest changing name and description to better reflect the nature of this example

sanjaypujare replied:
Done and commited.

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -0,0 +1,205 @@
+  private static final Logger LOG = LoggerFactory.getLogger(ApplicationTest.class);
+  
+  private static final String FILE_NAME = "test";
+  private static final String FILE_DIR  = "/tmp/FromAMQ";
```
Better to use relative path beginning with "target/" as done in some of the other examples (e.g. hdfs2kafka) since that is a known transient directory.

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -0,0 +1,205 @@
+      
+      createAMQClient();
+      
+      // write messages to SQS Queue                                                                                                            
```
"SQS Queue" => "ActiveMQ Queue"

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -0,0 +1,205 @@
+    conf.set(   pre + "filePath",        FILE_DIR);
+    conf.set(   pre + "baseName",        FILE_NAME);
+    conf.setInt(pre + "maxLength",       45);
+    conf.setInt(pre + "rotationWindows", 1);
```
Please set properties in an XML configuration file (e.g. properties.xml) and source that file here as done in several other examples, for instance: https://github.com/DataTorrent/examples/blob/master/tutorials/jdbcIngest/src/test/java/com/example/mydtapp/JdbcInputAppTest.java

Using properties files is preferred since it shows users how to change configuration values without having to change the code. You can override a couple here and add a comment to say that they can also be set in code, for illustrative purposes.

amberarrow commented:
- Please add a README.md describing what the example does, how to configure and run it.
- Add an entry to the table in the parent README.md (order is alphabetic).
- Test on cluster; if that requires some setup, describe what prerequisites need to be satisfied in order to run the example. If not too onerous, describe how to perform setup. Some of the JDBC examples have a similar need.
Rationale: People trying these examples are likely to be novices, so anything we can do to make the ride less bumpy will be appreciated !

amberarrow commented on tutorials/jmsActiveMQ/src/main/java/com/example/jmsActiveMQ/ActiveMQApplication.java:
```diff
@@ -0,0 +1,37 @@
+        new JMSStringInputOperator());
+
+    amqInput.setSubject(conf.get(QUEUE_NAME_PROPERTY));
+    amqInput.getConnectionFactoryProperties().put("brokerURL", "vm://localhost");
```
If possible, configure from properties file

amberarrow commented on tutorials/jmsActiveMQ/src/main/resources/META-INF/properties.xml:
```diff
@@ -0,0 +1,24 @@
+    <value>hello world: %s</value>
+  </property>
+</configuration>
+
```
Remove these and add configuration items for the current application.

amberarrow commented on tutorials/jmsActiveMQ/src/site/conf/my-app-conf1.xml:
```diff
@@ -0,0 +1,11 @@
+    <name>dt.application.MyFirstApplication.operator.randomGenerator.prop.numTuples</name>
+    <value>1000</value>
+  </property>
+</configuration>
```
Remove this item.

sanjaypujare replied:
I have kept the file but removed those other properties. I am using the file to illustrate for the example on the cluster.

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -0,0 +1,205 @@
+    // Create a Session
+    session = connection.createSession(false, Session.CLIENT_ACKNOWLEDGE);
+
+    // Create the destination (Topic or Queue)
```
"(Topic or Queue)" => queue

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -0,0 +1,205 @@
+    // Create the destination (Topic or Queue)
+    Destination destination = session.createQueue(QUEUE_NAME);
+
+    // Create a MessageProducer from the Session to the Topic or Queue
```
Suggest rewriting comment to be more descriptive, e.g.
Create producer for queue from session and set its delivery mode.

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -0,0 +1,205 @@
+    
+    // Clean up
+    session.close();
+    connection.close();
```
Suggest extracting setup and teardown to functions annotated with `@Before`, `@BeforeClass`, etc. as done in other tests.

sanjaypujare replied:
Considering the timing, it wouldn't make sense to move the clean up code to a separate @After function since the close needs to happen before the rest of the test code executes.

sanjaypujare commented:
@amberarrow  all of your comments have been incorporated. Pls review and merge.

amberarrow commented on tutorials/jmsActiveMQ/README.md:
```diff
@@ -0,0 +1,49 @@
+Unless you already have the ActiveMQ broker installed, download, install and
+start it as per http://activemq.apache.org/version-5-getting-started.html.
+For this example, let us assume the broker is installed on a machine with the IP address
+192.168.128.142. You can access the Admin UI as 
```
Minor point but would be good to eliminate trailing whitespace in a few lines in this file.

sanjaypujare replied:
Done.

amberarrow commented on tutorials/jmsActiveMQ/README.md:
```diff
@@ -0,0 +1,49 @@
+
+**Step 3**: Build the code:
+
+    shell> mvn clean install
```
Since there is no obvious benefit to installing, can we use: `mvn clean package -DskipTests`

sanjaypujare replied:
Done

amberarrow commented on tutorials/jmsActiveMQ/README.md:
```diff
@@ -0,0 +1,49 @@
+    shell> mvn clean install
+
+Upload the `target/jmsActiveMQ-1.0-SNAPSHOT.apa` to the UI console if available or launch it from
+the commandline using `apexcli`.
```
apexcli => apex

sanjaypujare replied:
Done

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -105,9 +111,15 @@ private void startJMSService() throws Exception
+    startEmbeddedActiveMQBroker();
     
     // Create a ConnectionFactory
     ActiveMQConnectionFactory connectionFactory = new ActiveMQConnectionFactory("vm://localhost");
```
Seems like the argument should come from conf.get("dt.operator.amqIn.prop.connectionFactoryProperties.brokerURL") ?

sanjaypujare replied:
Changed

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -65,10 +66,14 @@ public void testApplication() throws IOException, Exception {
       
+      getConfig();
+
+      qNameToUse = conf.get("dt.operator.amqIn.prop.subject");
```
Minor point: Both qNameToUse and the brokerURL could be retrieved in getConfig() and used whereever they are needed.

sanjaypujare replied:
Done

amberarrow commented on tutorials/jmsActiveMQ/src/test/java/com/example/jmsActiveMQ/ApplicationTest.java:
```diff
@@ -135,27 +147,30 @@ private void writeMsg(String[] msgs) throws JMSException
+  }
+
+  private void closeClient() throws JMSException {
+	  session.close();
```
TABs in file

sanjaypujare replied:
Fixed

amberarrow commented on tutorials/jmsActiveMQ/src/site/conf/my-app-conf1.xml:
```diff
@@ -1,11 +1,28 @@
-</configuration>
+	<property>
+		<name>dt.operator.fileOut.prop.filePath</name>
+		<value>target/</value>
```
Since this file is intended to be used on a cluster, shouldn't the value be a full  HDFS path like "/tmp" ? Does this value actually work ?

sanjaypujare replied:
Yes, it works. You will see this under /user/dtadmin/target (assuming the current Unix user is dtadmin and hence the current hadoop user is also dtadmin). I have left it as it is and added some explanation in the README file.

amberarrow commented:
Made a few more comments; please address those and rebase and squash
(there is a minor merge conflict in tutorials/README.md)

sanjaypujare commented:
@amberarrow  I have incorporated all/most of your review comments. Pls review and merge. Also I have resolved conflicts in tutorials/README.md so you will see those as changes from me.

amberarrow commented on tutorials/README.md:
```diff
@@ -1,17 +1,22 @@
 | hdht              | Shows how to use the HDHT operator. |
 | hdfs2kafka        | Shows how to read from HDFS and write to a Kafka topic. |
 | jdbcIngest        | Shows how to read rows from a table in an SQL database and write them to a file in HDFS |
+| jmsActiveMQ       | Shows how to use the JMS input operator to read from an ActiveMQ queue |
```
Needs to be moved down a line to maintain alphabetic order.

amberarrow commented on tutorials/jmsActiveMQ/pom.xml:
```diff
@@ -0,0 +1,288 @@
+    <dependency>
+      <groupId>org.apache.apex</groupId>
+      <artifactId>malhar-library</artifactId>
+      <version>3.5.0-SNAPSHOT</version>
```
3.5.0-SNAPSHOT => 3.5.0

amberarrow commented on tutorials/jmsActiveMQ/src/site/conf/my-app-conf1.xml:
```diff
@@ -0,0 +1,28 @@
+		<name>dt.operator.amqIn.prop.subject</name>
+		<value>test1</value>
+	</property>
+</configuration>
```
Missing final line terminator.
Also TABs present.

amberarrow commented:
Made a couple more minor comments.
I'm still getting conflicts in ../README.md when I try to merge; please rebase if necessary so it can be merged cleanly.

sanjaypujare commented:
@amberarrow I have incorporated all your comments and also resolved the conflict in tutorials/README.md, however git doesn't show it as resolved. So I have a branch new branch sanjaypujare-SPOI-8769.addition3 in which contains all the commits from this branch and doesn't have any conflicts. If you want we can close this PR and open another one sanjaypujare-SPOI-8769.addition3

amberarrow commented on tutorials/README.md:
```diff
@@ -11,12 +12,14 @@
 | hdfs2kafka        | Shows how to read from HDFS and write to a Kafka topic. |
+| jdbcIngest        | Shows how to read rows from a table in an SQL database, polling and non-polling fashion, and write them to a file in HDFS. |
+| innerjoin         | Shows how to use streaming innerjoin operator|
 | jdbcIngest        | Shows how to read rows from a table in an SQL database and write them to a file in HDFS |
```
Looks like we have 2 jdbcIngest lines of which the second is redundant and the first
needs to be after the innerjoin.

sanjaypujare replied:
Hmmm... this was from before my changes but I'll fix that. BTW the content of these lines are different but I'll remove the 2nd one as you suggest.

amberarrow commented:
After addressing the one comment I made above, if you rebase your changes on the current master and squash them, the conflicts should disappear.

sanjaypujare commented:
@amberarrow  I addressed your last comment and also rebased on the new master and squashed commits but the PR still shows a conflict in tutorials/README.md . Not sure what else I need to do :-(
